### PR TITLE
fix(curriculum): Reorder information in Record Collection challenge

### LINKED ISSUE #50182: Record Collection still could be clearer
### Describe the Issue

This part is still tripping people up

> If prop is tracks and value isn't an empty string, add the value to the end of the album's tracks array. You need to create this array first if the album does not have a tracks property.

### Affected Page

https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures/basic-javascript/record-collection

### Your code

Usually one piece or another gets lost:

```javascript
function updateRecords(records, id, prop, value) {
  if (value === "") {
     delete records[id][prop];
  } else if (prop !== "tracks" && value !== "") {
    records [id][prop] = value;
  } else if (prop === "tracks" &&  value !== "" && records[id].hasOwnProperty ("tracks") === false){
    records [id][prop] = [value];
  }
  return records;
} 
```

### Expected behavior

Maybe this is clearer

> If `prop` is `tracks` and `value` isn't an empty string, update the album's `tracks` array. You need to create this array first if the album does not have a `tracks` property. Then, add the `value` to the end of the album's `tracks` array.

Side note - I still don't love the fact that we say `tracks` like a variable instead of `"tracks"` like a string literal. I think that's confusing here. (I know, I know, the style guide hates "s)

### Screenshots

_No response_

### System

All

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jeremylt**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #50182

This focuses on the clarity of the instruction. I think we can handle this question of not being clear that `"tracks"`  is a string literal in a code block separately.

### CODE REVIEW

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md [hunk 1]
```diff
@@ -22,7 +22,7 @@ Complete the function using the rules below to modify the object passed to the f
 -   Your function must always return the entire `records` object.
 -   If `value` is an empty string, delete the given `prop` property from the album.
 -   If `prop` isn't `tracks` and `value` isn't an empty string, assign the `value` to that album's `prop`.
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md:24]:
```suggestion
-   If `value` is an empty string, delete the given `prop` property from the album.
-   If `prop` isn't the string `tracks` and `value` isn't an empty string, assign the `value` to that album's `prop`.
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md [hunk 2]
```diff
@@ -22,7 +22,7 @@ Complete the function using the rules below to modify the object passed to the f
...
--   If `prop` is `tracks` and value isn't an empty string, add the `value` to the end of the album's `tracks` array. You need to create this array first if the album does not have a `tracks` property.
+-   If `prop` is `tracks` and `value` isn't an empty string, update the album's `tracks` array. If the album does not have a `tracks` property, then you need to create this array first. Add the `value` to the end of the album's `tracks` array.
```
Comment (naomi-lgbt) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md:25]:
```suggestion
-   If `prop` is the string `tracks` and `value` isn't an empty string, update the album's `tracks` array. If the album does not have a `tracks` property, then you need to create this array first. Add the `value` to the end of the album's `tracks` array.
```

**Review by naomi-lgbt — COMMENTED** [review]
How would you feel about specifying this, while we're here?

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md (hunk 1 above)
Comment (jeremylt) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md:24]: That feels awkward. Its more literally accurate, but its also a mouthful

**Review by naomi-lgbt — APPROVED** [review]
Happy with this as is.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md (hunk 2 above)
Comment (lasjorg) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md:25]:
```suggestion
-   If `prop` is `tracks` and `value` isn't an empty string, add `value` to the end the album's `tracks` array. If the album does not have a `tracks` property, you must create and assign it an empty array. Then add the `value` to the album's `tracks` array.
```

> update the album's `tracks` array.

Seem a bit vague as to _where_ in the array `value` should go (start or end).

> Add the `value` to the end of the album's `tracks` array.

Adding the `value` to the end of the array only applies to the array when it exists, not when it is an empty array.

Comment (jeremylt) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md:25]: Adding the value to the end of an array means putting it after everything that's already there. You can certainly put the value after no elements. That's trivially easy and uses the exact same code as if there were elements in the array!

But that's my mathematical training that doesn't mind that in the least.

---

I keep trying to prevent  this bullet from feeling too wordy

Maybe
```suggestion
-   If `prop` is `tracks` and `value` isn't an empty string, you need to update the album's `tracks` array. First, if the album does not have a `tracks` property, assign it an empty array. Then add the `value` as the last item in the album's `tracks` array.
```

We need to communicate "to the end" somehow because order of the array contents matters in the tests.

Comment (lasjorg) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md:25]: I'm saying when the array exists and has elements in it the new value should be added to the end of the array (first paragraph, first requirement).

When the array does not exist and an empty array is created the position the value is added to is irrelevant as the array is empty (second paragraph, second requirement).

Comment (jeremylt) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md:25]: Right, and that ordering is what I am trying to fix. That's close to the current ordering, which produced nonsense code from a noticeable number of users. "Do this. But do this other thing if this subcondition is true" was too much for some learners to understand. The would only do one or the other, not read and consider all of the sentences together.

Comment (lasjorg) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md]: I just don't understand why the last sentence is where it is.

The way I read it, it connects to the previous sentence, which makes it sound like you are asking for the element to be added to an empty array as the last element.

Comment (jeremylt) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md]: Yup. That's what it means. That's a completely sensical instruction. You can add an element to an empty array:

```javascript
  } else if (prop === "tracks" && value !== "") {
    if (!records[id].hasOwnProperty("tracks")) {
      records[id]["tracks"] = [];
    }
    records[id]["tracks"].push(value);
  } else {
```

This is one of the guide solutions, more or less.

One way or another, at this point we are asking for replication of the pesudocode now given in this challenge. Too many of them just don't have the skills to think about what the words mean yet, unfortunately. So its a question of which pesudocode we want the users to replicate.

We could also 
```suggestion
-   If `prop` is `tracks` and `value` isn't an empty string and the album does not have a `tracks` property, assign a new array containing the `value` to the album's `tracks` array.
-   If `prop` is `tracks` and `value` isn't an empty string and the album does have a `tracks` property, add the `value` to the end of the album's `tracks` array.
```

If you want them to write

```javascript
  } else if (prop === "tracks" && value !== "" && !records[id].hasOwnProperty("tracks")) {
      records[id]["tracks"] = [ value];
  } else if (prop === "tracks" && value !== "" && records[id].hasOwnProperty("tracks"))
    records[id]["tracks"].push(value);
  } else {
```

I thought this was really wordy, but if you like it better go for it.

At this point, its bikeshedding for me to prefer mine, so commit and merge whichever you feel better meets our vision. I don't think it makes a huge difference.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md (hunk 1 above)
Comment (bbsmooth) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md:24]: Agreed, it's a little awkward perhaps, but isn't it a source of confusion?

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md (hunk 2 above)
Comment (lasjorg) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md]: > If the album does not have a `tracks` property, then you need to create this array first. Add the `value` to the end of the album's `tracks` array.

Last/end implies something comes before it, if it is an empty array how can the element be added to the end? What end? It is an empty array, there is nothing that comes before it. You might as well say the start and that would mean the same. There is no start or end because the array is empty.

---

Anyway, I won't harp on this anymore, I'm just saying I don't understand why it is worded like that.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md (hunk 1 above)
Comment (jeremylt) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.md:24]: I haven't noticed it as a particularly big issue on the forum. I've seen more confusion about where/how to check if the `track` property exists (though that could be confirmation bias?)

**Review by lasjorg — APPROVED** [review]
My comment is not blocking the merge so anyone can merge it. I will approve it and let someone else do the merge if no one sees an issue with it.

**Comment by naomi-lgbt** [comment]
Let's go ahead and get this in. I think this is a challenge we'll have to keep iterating over until we get it right - or until the new curriculum releases.
